pkg/apis/batch/validation/validation.go

```diff
@@ -486,9 +499,61 @@ func validatePodTemplateUpdate(spec, oldSpec batch.JobSpec, fldPath *field.Path,
 }
 
 // ValidateJobStatusUpdate validates an update to a JobStatus and returns an ErrorList with any errors.
-func ValidateJobStatusUpdate(status, oldStatus batch.JobStatus) field.ErrorList {
+func ValidateJobStatusUpdate(job, oldJob *batch.Job) field.ErrorList {
 	allErrs := field.ErrorList{}
-	allErrs = append(allErrs, validateJobStatus(&status, field.NewPath("status"))...)
+	statusFld := field.NewPath("status")
+	allErrs = append(allErrs, validateJobStatus(&job.Status, statusFld)...)
+
+	// A strengthened validation of the Job status transitions is needed since the
+	// Job managed-by label let's the Job object be controlled by external
+	// controllers. We want to make sure the transitions done by the external
+	// controllers meet the expectations of the clients of the Job API.
+	// For example, we verify that a Job in terminal state (Failed or Complete)
+	// does not flip to a non-terminal state.
+	// We verify the status fields only if changed or order to allow unrelated
+	// status updates to be possible.
+	if utilfeature.DefaultFeatureGate.Enabled(features.JobManagedByLabel) {
+		for _, cType := range []batch.JobConditionType{batch.JobFailed, batch.JobComplete, batch.JobFailureTarget} {
+			if isConditionTrue(oldJob.Status.Conditions, cType) && !isConditionTrue(job.Status.Conditions, cType) {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("conditions"), field.OmitValueType{}, fmt.Sprintf("cannot disable the terminal %s=True condition", string(cType))))
+			}
+		}
+		if isConditionTrue(job.Status.Conditions, batch.JobComplete) && isConditionTrue(job.Status.Conditions, batch.JobFailed) {
+			if !isConditionTrue(oldJob.Status.Conditions, batch.JobComplete) || !isConditionTrue(oldJob.Status.Conditions, batch.JobFailed) {
```

Comment:
can you please write a short comment to explain why we check the oldJob?

Reply:
Moved the part of the comment about this from the top here (to bring it closer to the checks with this property). Also rephrased. PTAL.

Reply:
looks better, thanks!